File: core/src/main/java/dev/sweety/core/exception/ExceptionUtils.java
package dev.sweety.core.exception;

import lombok.experimental.UtilityClass;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.function.Consumer;

@UtilityClass
public class ExceptionUtils {

    @SafeVarargs
    public <T> void throwIfAnyEquals(String message, T ifEquals, T... toCheck) {
        for (T o : toCheck) {
            if (o == ifEquals) throw new IllegalArgumentException(message);
        }
    }

    public static <T> T throwSilently(ThrowingSupplier<T> func, Consumer<Throwable> errorHandler) {
        try {
            return func.get();
        } catch (Throwable t) {
            errorHandler.accept(t);
            return null;
        }
    }

    public static <T> T throwSilently(ThrowingSupplier<T> func) {
        return throwSilently(func, k -> {});
    }

    @FunctionalInterface
    public interface ThrowingSupplier<T> {
        T get() throws Throwable;
    }

    public static String getStackTrace(final Throwable throwable) {
        if (throwable == null) return "";
        final StringWriter sw = new StringWriter();
        throwable.printStackTrace(new PrintWriter(sw, true));
        return sw.toString();
    }
}

File: util/exception/src/main/kotlin/dev/sweety/exception/ExceptionUtils.kt
package dev.sweety.exception

import java.io.PrintWriter
import java.io.StringWriter
import java.util.function.Consumer

object ExceptionUtils {

    @JvmStatic
    fun <T> throwIfAnyEquals(message: String, ifEquals: T, vararg toCheck: T) {
        for (o in toCheck) {
            if (o === ifEquals) throw IllegalArgumentException(message)
        }
    }

    @JvmStatic
    fun <T> throwSilently(func: ThrowingSupplier<T>, errorHandler: Consumer<Throwable>): T? {
        return try {
            func.get()
        } catch (t: Throwable) {
            errorHandler.accept(t)
            null
        }
    }

    @JvmStatic
    fun <T> throwSilently(func: ThrowingSupplier<T>): T? = throwSilently(func) { }

    fun interface ThrowingSupplier<T> {
        @Throws(Throwable::class)
        fun get(): T
    }

    @JvmStatic
    fun getStackTrace(throwable: Throwable?): String {
        if (throwable == null) return ""
        val sw = StringWriter()
        throwable.printStackTrace(PrintWriter(sw, true))
        return sw.toString()
    }
}
